
Singkatnya: Cross Validation adalah metode menguji kualitas model machine learning terutama untuk data yang sedikit

Kita tahu untuk melakukan evaluasi kita perlu membagi dataset menjadi dua yaitu training set dan test set. 
Cara pembagian dataset yang paling sederhana (primitif) biasanya dengan melakukan split 80 % training set dan  20 % test set.
Kalau kalian memiliki 10000 data , maka 8000 data digunakan sebagai training dan 2000 data sebagai testing

Lalu bagaimana kalau dataset terlalu kecil ?
Misal kita hanya memiliki data yang hanya berisi 50 sampel
Kalau memakai aturan 80:20 , maka kita hanya memiliki 40 sampel di training set dan 10 sampel di test set . Ini kurang untuk melihat kualitas sebenarnya dari model machine learning kita.

Dengan menggunakan Cross Validation maka kita bisa memiliki 50 training set dan 50 test set , bagaimana bisa ?

Cara yang dipakai dalam metode Cross Validation adalah dengan melakukan sampel berulang kali

Yang paling umum yaitu 10 fold Cross Validation

yaitu melakukan sample 10 kali , dengan ratio 90% training set dan 10 % test set

Dengan data berjumlah 50 maka :
Pada iterasi pertama akan diambil 40 training set , dan 10 test set

Pada iterasi kedua akan diambil 40 sampel untuk training set dan 10 sampel test set , namun 10 sampel test set yang dipakai kali ini berbeda dari 10 sampel test set yang sebelumnya.

Begitu seterusnya


Dengan cara ini 50 data yang ada bergantian terkadang menjadi training set , terkadang menjadi test set

Hasil prediksi lalu dgabung dari tiap iterasi , sehingga total menjadi 50 sampel , untuk dilihat kualitasnya melalui metrics yang ada , misal accuracy

Mari simak kode dibawah ini



In [5]:
from sklearn.model_selection import KFold
from sklearn.datasets import load_iris

In [14]:
dataset= load_iris() #ambil dataset contoh dari sklearn

X_dataset = dataset['data']
y_dataset = dataset['target']

   
    

mari kita hitung jumlah total data yang kita miliki

In [16]:
len(X_dataset)

150

ternyata kita memiliki 150 data, kalau kita melakukan split dengan cara primitif 80:20 , maka kita hanya memiliki data 120 utuk traning dan 30 untuk testing , 

Kalau menggunakan Cross Validation kita bisa memaksimalkan dataset kita

In [21]:
kfold = KFold(10) #kita memakai 10 fold Cross Validation
for train_index, test_index  in kfold.split(X_dataset):
    
    print("train_index: " , train_index)
    print("test_index: " , test_index)

train_index:  [ 15  16  17  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32
  33  34  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50
  51  52  53  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68
  69  70  71  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86
  87  88  89  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104
 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122
 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140
 141 142 143 144 145 146 147 148 149]
test_index:  [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14]
train_index:  [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  30  31  32
  33  34  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50
  51  52  53  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68
  69  70  71  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86
  87  88  89  90  91  92  93  94  95  96  97  98  99 100 101 102 103 10

Perhatikan , ada 10 kali perulangan , masing-masing perulangan memiliki test_index yang berbeda

In [22]:
kfold = KFold(10) #kita memakai 10 fold Cross Validation
for train_index, test_index  in kfold.split(X_dataset):
    
    print("train_index: " , len(train_index))
    print("test_index: " , len(test_index))

train_index:  135
test_index:  15
train_index:  135
test_index:  15
train_index:  135
test_index:  15
train_index:  135
test_index:  15
train_index:  135
test_index:  15
train_index:  135
test_index:  15
train_index:  135
test_index:  15
train_index:  135
test_index:  15
train_index:  135
test_index:  15
train_index:  135
test_index:  15


Perhatikan lagi , pada setiap perulangan/iterasi ada 135 data training yang digunakan, dan 15 data test yang digunakan.

data test pada setiap perulangan ini nantinya dijadikan satu. Sehingga total kita memiliki 150 data test ( karena ada 10 perulangan)


Coba bandingkan dengan melakukan split dengan cara primitif 80:20 tadi dimana kita hanya memiliki 30 data test. Dengan Cross Validation kita memiliki 150 data test. Dengan lebih banyak data test maka hasil metrics lebih solid. 
    Kita bisa menghitung accuracy, precision, recall, f-measure dengan hasil yang lebih dapat dipercaya karena data test lebih besar